<div class="alert alert-block alert-info"><b></b>
<h1><center> <font color='black'> Homework 01  </font></center></h1>
<h2><center> <font color='black'> Introduction and first look at the data </font></center></h2>    
<h2><center> <font color='black'> BDA - University of Tartu - Spring 2020</font></center></h3>
</div>

# 1. Python and Pandas (4.5 points)

* In this course we are going to use Python, and mainly Pandas data structures and data analysis tool. please make sure that you have went through the practise session carefully. After that please answer the following questions below about the abalone dataset.

First, let's import pandas package as pd and read data from the CSV file into a pandas DataFrame.

In [44]:
import pandas as pd
data = pd.read_csv("abalone.csv")
data

Gender  Length  Diameter  Height  Weight  Rings
0        F   0.505     0.385   0.135  0.6185     12
1        F   0.650     0.475   0.165  1.3875      9
2        I   0.520     0.380   0.135  0.5395      8
3        F   0.550     0.425   0.140  0.9520      7
4        M   0.640     0.525   0.180  1.3135     10
..     ...     ...       ...     ...     ...    ...
995      I   0.550     0.415   0.135  0.8095     12
996      I   0.455     0.435   0.110  0.4265      8
997      M   0.660     0.500   0.165  1.3195      9
998      I   0.525     0.400   0.130  0.6455      8
999      M   0.515     0.395   0.135  1.0070      8

[1000 rows x 6 columns]

`Question 0` is an example of how you should answer the following questions. First, in the `#TODO` cell you have to write the code representing your solution. And in the answer cell write the answer with your own words. It does not have to be a full sentence.

 **1.0. What is overall memory consumption of the data?** (EXAMPLE)

In [3]:
# TODO - here you have to write the code how did you find the answer.
data.memory_usage(deep=True).sum()

106128

**<font color='red'>Answer:</font>** The total memory usage is 102080 bytes (102,08 kb).

 **1.1. What are the column names of the dataset?** **(0.25 point)**

In [7]:
# TODO
data.columns

Index(['Gender', 'Length', 'Diameter', 'Height', 'Weight', 'Rings'], dtype='object')

**<font color='red'>Answer:</font>** 'Gender', 'Length', 'Diameter', 'Height', 'Weight', 'Rings'

**1.2. How many observations (i.e. rows) are in this data frame?** **(0.25 point)**

In [11]:
# TODO
data.index

RangeIndex(start=0, stop=1000, step=1)

**<font color='red'>Answer:</font>** 1000

**1.3. Print the first 5 lines from the dataset. What are the values of feature "Rings" of the printed observations?** **(0.25 point)**

In [14]:
# TODO
data.head()

Gender  Length  Diameter  Height  Weight  Rings
0      F   0.505     0.385   0.135  0.6185     12
1      F   0.650     0.475   0.165  1.3875      9
2      I   0.520     0.380   0.135  0.5395      8
3      F   0.550     0.425   0.140  0.9520      7
4      M   0.640     0.525   0.180  1.3135     10

**<font color='red'>Answer:</font>** 12, 9, 8, 7, 10

**1.4. Extract the last 3 rows of the data frame. What is the "Weight" of these abalones?** **(0.25 point)**

In [16]:
# TODO
data['Weight'].tail(3)

997    1.3195
998    0.6455
999    1.0070
Name: Weight, dtype: float64

**<font color='red'>Answer:</font>** 1.3195, 0.6455, 1.0070

**1.5. What is the value of diameter in the row 577?** **(1 point)**

In [20]:
# TODO
data.loc[576]['Diameter']

0.51

**<font color='red'>Answer:</font>** 0.51

**1.6. What is the mean of the height column?** **(0.5 point)** <br>
Hint 1: Use pandas describe() method

In [24]:
# TODO
data['Height'].describe()

count    1000.000000
mean        0.141100
std         0.057064
min         0.000000
25%         0.115000
50%         0.145000
75%         0.165000
max         1.130000
Name: Height, dtype: float64

**<font color='red'>Answer:</font>** 0.141100

**1.7. Extract the subset of rows of the data frame where gender is M and weight values are below 0.75. What is the mean of diameter in this subset?** **(1 point)**

In [45]:
# TODO
data[(data['Gender'] == "M") & (data['Weight'] < 0.75)].describe()

Length    Diameter      Height      Weight       Rings
count  119.000000  119.000000  119.000000  119.000000  119.000000
mean     0.444118    0.342647    0.115504    0.467366    8.865546
std      0.075819    0.063027    0.022193    0.196004    1.986925
min      0.235000    0.160000    0.055000    0.051500    4.000000
25%      0.385000    0.295000    0.100000    0.291500    7.500000
50%      0.465000    0.360000    0.115000    0.495500    9.000000
75%      0.500000    0.385000    0.130000    0.643250   10.000000
max      0.575000    0.455000    0.190000    0.749500   14.000000

**<font color='red'>Answer:</font>** 0.342647

**1.8. What is the minimum of length when rings is equal to 18?** **(1 point)**

In [53]:
# TODO
data[data['Rings'] == 18].describe()

Length   Diameter     Height     Weight  Rings
count  12.000000  12.000000  12.000000  12.000000   12.0
mean    0.570417   0.451250   0.172917   1.076625   18.0
std     0.074084   0.058742   0.028956   0.414502    0.0
min     0.465000   0.350000   0.120000   0.494000   18.0
25%     0.523750   0.408750   0.162500   0.872125   18.0
50%     0.557500   0.442500   0.177500   1.004750   18.0
75%     0.608750   0.481250   0.195000   1.199500   18.0
max     0.725000   0.560000   0.210000   2.141000   18.0

**<font color='red'>Answer:</font>** 0.465000

# 2.  Data understanding (2.5 point)

Collecting, describing, exploring and verifying data quality are <b>very important</b> steps in any data science task. In this exercise the main goal is to understand the attributes in the 'adult' dataset.<br>

**2.0. Load data from the  `adult.csv` file.** <br>
Hint 1: The extension of the filename may not always reflect the structure of data precisely. Are the values comma-separated (csv) or not? 

In [0]:
import numpy as np
import pandas as pd

# TODO
data = ...

**2.1. Name all the columns in this dataset and describe the meaning of each column with one sentence.** <br>

In [0]:
# TODO

**<font color='red'>Answer:</font>** Answer goes here.

####  2.1.1. In the following subtasks, you will have to visualise the distribution of the values for different features using appropriate diagrams. Visualisation is one of the easiest ways to get a good overview of the data and to notice any oddities, outliers etc.

In [0]:
# Matplotlib is a Python 2D plotting library which produces publication quality figures in a variety of hardcopy formats 
# and interactive environments across platforms. Matplotlib can be used in Python scripts, the Python and IPython shells, 
# the Jupyter notebook, web application servers, and four graphical user interface toolkits.
import matplotlib.pyplot as plt

# allows to output plots in the notebook
%matplotlib inline 

# Set the default style
plt.style.use("ggplot") 

As an example, here is the age frequency histogram, where age is an attribute of type **numeric**:

In [0]:
# Columns may contain non-numeric values, errors or missing values. Therefore, non-numeric values must be dealt with.
if data is not ...:
    pd.to_numeric(data['age'], errors='coerce').hist(bins=30) # ‘coerce’ -> invalid parsing will be set as NaN
else:
    print('Please define `data` in earlier subtasks')

**2.2. Attribute `Workclass`. First, specify the type of the attribute (nominal / ordinal / numeric) and second, choose an appropriate chart type and visualise the distribution in the relative frequency scale.**

**<font color='red'>Attribute type :</font>** Answer goes here.

In [0]:
# TODO

**2.3. Attribute `Education`. First, specify the type of the attribute (nominal / ordinal / numeric) and second, choose an appropriate chart type and visualise the distribution in the relative frequency scale.**

**<font color='red'>Attribute type :</font>** Answer goes here.

In [0]:
# TODO

**2.4. Attribute `Occupation`. First, specify the type of the attribute (nominal / ordinal / numeric) and second, choose an appropriate chart type and visualise the distribution in the relative frequency scale.**

**<font color='red'>Attribute type :</font>** Answer goes here.

In [0]:
# TODO

**2.5. Attribute `Native Country`. First, specify the type of the attribute (nominal / ordinal / numeric) and second, choose an appropriate chart type and visualise the distribution in the relative frequency scale.**

**<font color='red'>Attribute type :</font>** Answer goes here.

In [0]:
# TODO

**2.6. Attribute `Salaries`. First, specify the type of the attribute (nominal / ordinal / numeric) and second, choose an appropriate chart type and visualise the distribution in the density scale.**

**<font color='red'>Attribute type :</font>** Answer goes here.

In [0]:
# TODO 

**2.7. Attributes `Male` and `Female`. First, specify the types of these attributes (nominal / ordinal / numeric) and second, choose an appropriate chart type and visualise the relative frequency of positive values in attribute `Male` and positive values in attribute `Female` (relative frequency of males and females).**

**<font color='red'>Attribute type :</font>** Answer goes here.

In [0]:
# TODO - Put Male and Female attributes on the same plot for comparison

**2.8. Count the number of missing values in each attribute in the dataset and also report the percentage of missing values out of all rows (the original file uses '?' for missing values)**

In [0]:
#TODO

total_rows = ...
if data is not ...:
    for column in data.columns:
        total_missing = ...
        if total_missing > 0:
            print (...)
else:
    print('Please define `data` in earlier subtasks')

# 3.  Data preparation (1.5 points)

Preparing the data for analysis is the most crucial phase in data analysis. In this exercise, the main goal is to clean up the dataset 'adult'. The dataset has been made unclean on purpose so that you can practice cleaning it up. Most of the introduced uncleanliness is of the same kinds as described in slides of Lecture 02. Additionally, some values have been introduced which are non-sensical due to the meaning of the attribute.<br>

First, let's make a deep copy of the original DataFrame. Later we will compare it to the cleaned/modified DataFrame.

In [0]:
if data is not ...:
    original_data = data.copy(deep=True) # Make a deep copy, including a copy of the data and the indices

**3.1. Sometimes values include leading and/or trailing spaces. Remove leading and trailing whitespace from strings. Count how many rows were changed in the dataset.** <br>

In [0]:
# TODO

Use code below to count how many rows were affected.

In [0]:
# Count differences
if data is not ...:
    data_all = pd.concat([original_data, data]).drop_duplicates()
    diff = data_all.shape[0] - data.shape[0]
    print ('Difference: ' + str(diff))

**<font color='red'>Answer:</font>** Answer goes here.

**3.2. Replace all non-sensical values in `data` with np.nan from numpy package to denote missing values (Original file uses '?' for missing values). Fix the typos in the dataset. Justify all the changes that you do, explaining why you are sure that this is a correct change and what you think the cause for such error could be.**  <br>

In [0]:
# TODO

**<font color='red'>Answer:</font>** Answer goes here.

**3.3. Count the rows which were changed in the subtask b** <br>

In [0]:
# TODO

**<font color='red'>Answer:</font>** Answer goes here.

**3.4. Report values that look suspicious whereas you are not sure whether they are definitely wrong** <br> 

**<font color='red'>Answer:</font>** Answer goes here.

# 4.  Gathering interesting facts about the adult dataset (1.5 points)

Useful information about the data can be acquired through visualization even before modeling.

**4.1. Convert the attribute `Education` into type `category` and introduce the order for this attribute based on the educational level (e.g. Masters is higher education than Bachelors; use your best guess in ordering those levels for which the order is not obvious). Create a bar chart visualising the relative frequencies of values.**

In [0]:
# TODO

**4.2.  Create and print a Pandas data frame where each row stands for an occupation, each column stands for a level of education, and the cells in the table contain the average salary of people with the corresponding occupation and education level.** <br> 
Hint 1: You can use crosstab, see also aggfunc from the pandas documentation.

In [0]:
# TODO
crosstab_df = ...

**4.3. Make a separate plot for each row (occupation) of the table created in subtask b. Choose the type of the plot to convey the information as well as you can.** <br>

In [0]:
if crosstab_df is not ...:
    for index, row in crosstab_df.iterrows():
        row.plot(...) # TODO 
        plt.show()

**4.4. List 3 interesting facts that you can read out of these plots from subtask c.** <br>

**<font color='red'>Fact 1:</font>** Answer goes here.

**<font color='red'>Fact 2:</font>** Answer goes here.

**<font color='red'>Fact 3:</font>** Answer goes here.

## How long did it take you to solve the homework?

* Please answer as precisely as you can. It does not affect your points or grade in any way. It is okey, if it took 0.5 hours or 24 hours. The collected information will be used to improve future homeworks.

<font color='red'> **Answer:**</font>

**<font color='red'>(please change X in the next cell into your estimate)</font>**

X hours

## How much the homework is difficult?
you can put only number between $0:10$ ($0:$ easy, $10:$ difficult)

<font color='red'> **Answer:**</font>


